enhancements/network/smart-nic-ovn-offload.md

```diff
@@ -0,0 +1,456 @@
+| |                      +-+------+---->   K8s APIServer     |   |       |        SmartNIC          |
+| |  full mode           | |      |    |                     |   |       |                          |
+| +----------------------+ |      |    +---------^-----------+   |       | +----------------------+ |
+|                          |      |              |               |       | |  ovn-controller      | |
++------------+-------------+      +--------------+---------------+       | +----------------------+ |
+             |                                   |                       |                          |
+             |                                   |                       | +----------------------+ |
++------------+-------------+                     |                       | |  ovnkube-node in     | |
+|       Non-SmartNIC       |                     +-----------------------+-+                      | |
++--------------------------+                                             | |  smart-nic mode      | |
+                                                                         | +----------------------+ |
+                                                                         |                          |
+                                                                         +--------------------------+
+```
+
+On SmartNIC ARM cores, ovn-controller and ovnkube-node are running as
+containers, OvS is running as system service.
+On SmartNIC node, ovnkube-node is running as a container.
+On non-SmartNIC node, ovn-controller and ovnkube-node are running as
+containers, OvS is running as system service.
+
+#### Cluster Network Operator
+
+A new label `network.operator.openshift.io/external-ovn` will be introduced
+in Cluster Network Operator (CNO) to allow proper scheduling of OVN-Kubernetes
+components on SmartNIC nodes and non-SmartNIC nodes.
+
+Ovnkube-node-smart-nic daemonset will be added in CNO ovn-kubernetes manifests,
+which executes ovnkube-node in `smart-nic-host` mode. ovnkube-node-smart-nic
+daemonset runs only on SmartNIC nodes.
+
+Ovnkube-node-smart-nic pod will be mounted with the config map
+`ovn-management-ports` to get the ovn management port information of the node
+it is running on. The ovn management port information is provided by SR-IOV
+Network Operator and is used as argument to start ovnkube-node in
+`smart-nic-host` mode. ovnkube-node-smart-nic pod waits until the
+`ovn-management-ports` config map is created, then starts. The management
+port is an SR-IOV VF device that will be configured to `ovn-k8s-mp0` port by
+ovnkube-node.
+
+Daemonsets of OVN-Kubernetes components other than ovnkube-node-smart-nic will
+be modified with `nodeAffinity` to run on non-SmartNIC nodes.
+
+#### SR-IOV Network Operator
+
+SmartNIC offload will be enabled for nodes in custom MachineConfigPool.
+
+A new field `OvnHardwareOffload` will be added in `SriovOperatorConfig` CRD
```

Comment:
> It seems wrong for OVN offload to be configured in the SR-IOV operator config...

I agree with you on where the OVN offload shall be configured.

Some backgroup of why it is added in SR-IOV:
There was similar discussion when adding OVS HWOL (w/ CX-5) support: https://github.com/openshift/cluster-network-operator/pull/843, whether CNO or SR-IOV is the right place to expose user-facing APIs. The result is SR-IOV is used.  The previous point I think was we don't have example exposing pool specific config in CNO since all configs in CNO are applied to the entire cluster nodes and only nodes with SR-IOV capable device can be used for offload (so it requires SR-IOV be deployed on that nodes).

OVN offload is different from OVS offload (at least with current design), at that a node (w/ SmartNIC) is turned to a SmartNIC host which can no longer hold regular veth pods.

With the agreement that SmartNIC nodes will be managed in infrastructure management cluster (separate from the tenant cluster), I think it now becomes a cluster-level config instead of a node or device level thing. so having the user-facing API in CNO might be the proper choice (because we may not need to run SR-IOV in infra management cluster), also there might be use case that we don't need SR-IOV for ovn offload.

> The real feature here is "OVN offload". SR-IOV is merely an implementation detail. You should configure OVN offload in the CNO config, and then either CNO should make changes to the SR-IOV config as needed, or else the SR-IOV operator should read the CNO config.
> 
> (I guess this is related to my annoyance with needing SR-IOV annotations; I feel like the use of SR-IOV here should be completely hidden from the user. It's not like in the normal SR-IOV use case where you have a node where some pods are using SR-IOV and some pods aren't, so it makes sense that the user has to explicitly indicate which pods they want to use SR-IOV with. In the OVN offload cases, the user should be indicating "I want this pod to run on a node with OVN offload", and then that should be it. They shouldn't have to also explain to kubernetes how OVN offload works...)

There is another consideration in using SR-IOV Operator, that not all OVN offload configuration are done in single component like CNO, there needs coordination between several openshift operators, such as DNS, MCO, CNO, SR-IOV.  I'm not sure if it's a good idea to build awareness of other OpenShift Operator APIs in CNO since it will introduce dependency and complexity when doing cluster upgrade.